components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -5,68 +5,565 @@
 
 #include "brave/components/brave_news/browser/feed_v2_builder.h"
 
+#include <stddef.h>
+#include <algorithm>
+#include <cmath>
 #include <iterator>
-#include <locale>
+#include <string>
+#include <tuple>
 #include <utility>
 #include <vector>
 
+#include "base/containers/contains.h"
+#include "base/containers/flat_set.h"
 #include "base/functional/bind.h"
+#include "base/logging.h"
 #include "base/memory/scoped_refptr.h"
+#include "base/rand_util.h"
 #include "base/ranges/algorithm.h"
+#include "base/time/time.h"
 #include "brave/components/brave_news/browser/channels_controller.h"
 #include "brave/components/brave_news/browser/feed_fetcher.h"
 #include "brave/components/brave_news/browser/publishers_controller.h"
-#include "brave/components/brave_news/common/brave_news.mojom-forward.h"
+#include "brave/components/brave_news/browser/signal_calculator.h"
 #include "brave/components/brave_news/common/brave_news.mojom.h"
+#include "components/prefs/pref_service.h"
 #include "services/network/public/cpp/shared_url_loader_factory.h"
 
 namespace brave_news {
 
+namespace {
+
+// Minimum number of articles in an inline block.
+constexpr int kBlockInlineMin = 1;
+
+// Maximum number of articles in an inline block.
+constexpr int kBlockInlineMax = 5;
+
+// Ratio of inline articles to discovery articles.
+constexpr double kInlineDiscoveryRatio = 0.25;
+
+// Weighting for unsubscribed sources. Small but non-zero, so we still include
+// them in the feed.
+constexpr double kSourceSubscribedMin = 1e-5;
+
+// Weighting for subscribed sources (either this or |kSourceSubscribedMin| will
+// be applied).
+constexpr double kSourceSubscribedMax = 1;
+
+// Multiplier for unvisited sources. |visit_weighting| is in the range [0, 1]
+// inclusive and will be shifted by this (i.e. [0, 1] ==> [0.2, 1.2]). This
+// ensures we still show unvisited sources in the feed.
+constexpr double kSourceVisitsMin = 0.2;
+
+// Every N hours the popRecency half life will halve. I.e, if this was 24, every
+// day the popularity score will be halved.
+constexpr double kPopRecencyHalfLifeInHours = 18;
```

Comment:
It seems most of these constants are used only one places. Could you move them into the functions where they're used? I think limiting the visible scope would be better.

Reply:
Can do - I originally added them here because we have an end goal of moving these into prefs, so that we can play around with configuring them